libraries/chain/db_block.cpp

```diff
@@ -331,10 +331,19 @@ signed_block database::_generate_block(
    witness_id_type scheduled_witness = get_scheduled_witness( slot_num );
    FC_ASSERT( scheduled_witness == witness_id );
 
-   const auto& witness_obj = witness_id(*this);
-
    if( !(skip & skip_witness_signature) )
-      FC_ASSERT( witness_obj.signing_key == block_signing_private_key.get_public_key() );
+   {
+      auto signing_key = find_witness_key_from_cache( witness_id );
```

Comment:
Instead of using the cache, simply move the existing FC_ASSERT down a couple of lines, after ``_pending_tx_session.reset();``

Reply:
It's cheaper to check here than checking after called `_pending_tx_session.reset();`.

Reply:
Even for nodes with proper keys, it's still slightly cheaper because the cache would be much smaller than object database.

Reply:
For nodes with keys it may be cheaper, but the global cache is a penalty for *all* nodes.